File: packages/shared-ts/src/logging/format.ts
import { normalizeMeta } from './meta';
import { sanitizeValue } from './sanitize';
import type { LogLevel } from './structured-logger';

interface LogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  meta?: unknown;
  environment?: string;
  nodeVersion?: string;
  correlationId?: string;
}

export const formatLogEntry = (params: {
  level: LogLevel;
  message: string;
  meta: unknown;
  environment: string;
  nodeVersion: string;
  correlationId: string | undefined;
  baseMeta: Record<string, unknown>;
  getLogMetadata: () => Record<string, unknown>;
}): string => {
  const sanitizedMessage = String(sanitizeValue(params.message));

  const entry: LogEntry = {
    timestamp: new Date().toISOString(),
    level: params.level,
    message: sanitizedMessage,
    environment: params.environment,
    nodeVersion: params.nodeVersion,
  };

  if (params.correlationId) {
    entry.correlationId = params.correlationId;
  }

  const normalized = normalizeMeta(params.baseMeta, params.getLogMetadata, params.meta);
  if (normalized) {
    entry.meta = sanitizeValue(normalized);
  }

  return JSON.stringify(entry);
};

File: packages/adapters/pkg/logging/format.go
package logging

import (
	"encoding/json"
	"runtime"
	"time"
)

// LogEntry represents the structured JSON log format.
type LogEntry struct {
	Timestamp     string   `json:"timestamp"`
	Level         LogLevel `json:"level"`
	Message       string   `json:"message"`
	Meta          any      `json:"meta,omitempty"`
	Environment   string   `json:"environment,omitempty"`
	GoVersion     string   `json:"goVersion,omitempty"`
	CorrelationID string   `json:"correlationId,omitempty"`
}

// marshalFunc is the function used to marshal log entries. Can be overridden in tests.
var marshalFunc = json.Marshal

// FormatLogEntry produces a JSON string representing a log entry.
func FormatLogEntry(
	level LogLevel,
	message string,
	meta any,
	environment string,
	correlationID string,
	baseMeta map[string]any,
	contextMeta map[string]any,
) string {
	sanitizedMessage := ""
	if s, ok := SanitizeValue(message).(string); ok {
		sanitizedMessage = s
	}

	entry := LogEntry{
		Timestamp:     time.Now().UTC().Format(time.RFC3339),
		Level:         level,
		Message:       sanitizedMessage,
		Environment:   environment,
		GoVersion:     runtime.Version(),
		CorrelationID: correlationID,
	}

	normalized := NormalizeMeta(baseMeta, contextMeta, meta)
	if normalized != nil {
		entry.Meta = SanitizeValue(normalized)
	}

	data, err := marshalFunc(entry)
	if err != nil {
		// Fallback for failed marshaling
		return `{"level":"` + string(level) + `","message":"error marshaling log entry"}`
	}
	return string(data)
}
